adapters/lora-vertx/src/main/java/org/eclipse/hono/adapter/lora/providers/FireflyProvider.java

```diff
@@ -76,6 +86,43 @@ public LoraMessageType extractMessageType(final JsonObject loraMessage) {
         if (loraMessage.containsKey(FIELD_FIREFLY_FUNCTION_PORT)) {
             returnMap.put(LoraConstants.APP_PROPERTY_FUNCTION_PORT, loraMessage.getInteger(FIELD_FIREFLY_FUNCTION_PORT));
         }
+        final JsonObject dataRateJson = loraMessage.getJsonObject(FIELD_FIREFLY_SERVER_DATA, new JsonObject());
+
+        if (dataRateJson.containsKey(FIELD_FIREFLY_DATA_RATE)){
+            returnMap.put(LoraConstants.DATA_RATE, dataRateJson.getString(FIELD_FIREFLY_DATA_RATE));
+        }
+        if (dataRateJson.containsKey(FIELD_FIREFLY_CODING_RATE)){
+            returnMap.put(LoraConstants.CODING_RATE, dataRateJson.getString(FIELD_FIREFLY_CODING_RATE));
+        }
+        if (dataRateJson.containsKey(FIELD_FIREFLY_FREQUENCY)){
+            returnMap.put(LoraConstants.FREQUENCY, dataRateJson.getDouble(FIELD_FIREFLY_FREQUENCY));
+        }
+
+        final JsonObject parsedPacketJson = loraMessage.getJsonObject(FIELD_FIREFLY_PARSED_PACKET, new JsonObject());
+        if (parsedPacketJson.containsKey(FIELD_FIREFLY_FRAME_COUNT)){
+            returnMap.put(LoraConstants.FRAME_COUNT, parsedPacketJson.getInteger(FIELD_FIREFLY_FRAME_COUNT));
+        }
+
+        if (dataRateJson.containsKey(FIELD_FIREFLY_GATEWAY_RX)){
+            final JsonArray gwRxs = dataRateJson.getJsonArray(FIELD_FIREFLY_GATEWAY_RX);
+            final JsonArray normalizedGatways = new JsonArray();
+            for (int i = 0; i < gwRxs.size(); i++){
```

Comment:
missing whitespace between `)` and `{`

Reply:
Done.
For my understanding: this is required after a for loop but not after an "if"?

Reply:
it is always required, I just didn't notice that you omitted them after the `if` condition as well ...

Reply:
ok, changed the `if`s as well.